// discordx/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  ApplicationCommandMixin: () => ApplicationCommandMixin,
  Bot: () => Bot,
  ButtonComponent: () => ButtonComponent,
  Client: () => Client,
  ComponentType: () => ComponentType,
  ContextMenu: () => ContextMenu,
  DApplicationCommand: () => DApplicationCommand,
  DApplicationCommandGroup: () => DApplicationCommandGroup,
  DApplicationCommandOption: () => DApplicationCommandOption,
  DApplicationCommandOptionChoice: () => DApplicationCommandOptionChoice,
  DComponent: () => DComponent,
  DDiscord: () => DDiscord,
  DGuard: () => DGuard,
  DOn: () => DOn,
  DReaction: () => DReaction,
  DSimpleCommand: () => DSimpleCommand,
  DSimpleCommandOption: () => DSimpleCommandOption,
  Discord: () => Discord,
  Guard: () => Guard,
  Guild: () => Guild,
  MetadataStorage: () => MetadataStorage,
  ModalComponent: () => ModalComponent,
  On: () => On,
  Once: () => Once,
  Reaction: () => Reaction,
  RecursivelyMatchField: () => RecursivelyMatchField,
  SelectMenuComponent: () => SelectMenuComponent,
  SimpleCommand: () => SimpleCommand,
  SimpleCommandMessage: () => SimpleCommandMessage,
  SimpleCommandOption: () => SimpleCommandOption,
  SimpleCommandOptionType: () => SimpleCommandOptionType,
  SimpleCommandParseType: () => SimpleCommandParseType,
  Slash: () => Slash,
  SlashChoice: () => SlashChoice,
  SlashGroup: () => SlashGroup,
  SlashNameValidator: () => SlashNameValidator,
  SlashOption: () => SlashOption,
  SpecialCharactersList: () => SpecialCharactersList,
  isApplicationCommandEqual: () => isApplicationCommandEqual,
  resolveIGuilds: () => resolveIGuilds,
  toStringArray: () => toStringArray
});
module.exports = __toCommonJS(index_exports);

// src/classes/Mixin/ApplicationCommandMixin.ts
var ApplicationCommandMixin = class {
  constructor(command, instance) {
    this.command = command;
    this.instance = instance;
  }
  get name() {
    return this.command.name;
  }
  get description() {
    return this.command.description;
  }
};

// src/classes/SimpleCommandMessage.ts
var import_crypto = __toESM(require("crypto"));
var import_discord = require("discord.js");
var SimpleCommandMessage = class {
  constructor(prefix, argString, message, info, splitter) {
    this.prefix = prefix;
    this.argString = argString;
    this.message = message;
    this.info = info;
    this.splitter = splitter;
  }
  options = [];
  get name() {
    return this.info.name;
  }
  get description() {
    return this.info.description;
  }
  /**
   * Resolve options
   */
  resolveOptions() {
    return this.info.parseParamsEx(this);
  }
  /**
   * Verify that all options are valid
   *
   * @returns
   */
  isValid() {
    return !this.options.includes(null);
  }
  /**
   * Get related commands
   *
   * @returns
   */
  getRelatedCommands() {
    const commandName = this.info.name.split(" ")[0];
    if (!commandName) {
      return [];
    }
    return MetadataStorage.instance.simpleCommands.filter(
      (cmd) => cmd.name.startsWith(commandName) && cmd.name !== this.info.name
    );
  }
  /**
   * Send usage syntax for command
   *
   * @returns
   */
  sendUsageSyntax() {
    const maxLength = !this.info.options.length ? 0 : this.info.options.reduce(
      (a, b) => a.name.length > b.name.length ? a : b
    ).name.length;
    const embed = new import_discord.EmbedBuilder();
    embed.setColor(import_crypto.default.randomInt(654321));
    embed.setTitle("Command Info");
    embed.addFields({ name: "Name", value: this.info.name });
    embed.addFields({ name: "Description", value: this.info.description });
    if (this.info.aliases.length) {
      embed.addFields({ name: "Aliases", value: this.info.aliases.join(", ") });
    }
    embed.addFields({
      name: "Command Usage",
      value: `\`\`\`${this.prefix.toString()}${this.name} ${this.info.options.map((op) => `{${op.name}: ${SimpleCommandOptionType[op.type]}}`).join(" ")}\`\`\``
    });
    if (this.info.options.length) {
      embed.addFields({
        name: "Options",
        value: `\`\`\`${this.info.options.map((op) => `${op.name.padEnd(maxLength + 2)}: ${op.description}`).join("\n")}\`\`\``
      });
    }
    return this.message.reply({ embeds: [embed] });
  }
};

// src/Client.ts
var import_di = require("@discordx/di");
var import_discord2 = require("discord.js");
var import_escapeRegExp = __toESM(require("lodash/escapeRegExp.js"));
var Client = class _Client extends import_discord2.Client {
  _botId;
  _isBuilt = false;
  _prefix;
  _simpleCommandConfig;
  _silent;
  _listeners = /* @__PURE__ */ new Map();
  _botGuilds = [];
  _guards = [];
  logger;
  // static getters
  static get applicationCommandSlashesFlat() {
    return MetadataStorage.instance.applicationCommandSlashesFlat;
  }
  static get applicationCommandSlashes() {
    return MetadataStorage.instance.applicationCommandSlashes;
  }
  static get applicationCommandUsers() {
    return MetadataStorage.instance.applicationCommandUsers;
  }
  static get applicationCommandMessages() {
    return MetadataStorage.instance.applicationCommandMessages;
  }
  static get applicationCommandSlashOptions() {
    return MetadataStorage.instance.applicationCommandSlashOptions;
  }
  static get applicationCommands() {
    return MetadataStorage.instance.applicationCommands;
  }
  static get applicationCommandSlashGroups() {
    return MetadataStorage.instance.applicationCommandSlashGroups;
  }
  static get applicationCommandSlashSubGroups() {
    return MetadataStorage.instance.applicationCommandSlashSubGroups;
  }
  static get buttonComponents() {
    return MetadataStorage.instance.buttonComponents;
  }
  static get discords() {
    return MetadataStorage.instance.discords;
  }
  static get events() {
    return MetadataStorage.instance.events;
  }
  static get instance() {
    return MetadataStorage.instance;
  }
  static get modalComponents() {
    return MetadataStorage.instance.modalComponents;
  }
  static get reactions() {
    return MetadataStorage.instance.reactions;
  }
  static get selectMenuComponents() {
    return MetadataStorage.instance.selectMenuComponents;
  }
  static get simpleCommandsByName() {
    return MetadataStorage.instance.simpleCommandsByName;
  }
  static get simpleCommandMappedPrefix() {
    return MetadataStorage.instance.simpleCommandMappedPrefix;
  }
  static get simpleCommands() {
    return MetadataStorage.instance.simpleCommands;
  }
  // map static getters
  get applicationCommandSlashes() {
    return _Client.applicationCommandSlashes;
  }
  get applicationCommandSlashesFlat() {
    return _Client.applicationCommandSlashesFlat;
  }
  get applicationCommandSlashOptions() {
    return _Client.applicationCommandSlashOptions;
  }
  get applicationCommandSlashGroups() {
    return _Client.applicationCommandSlashGroups;
  }
  get applicationCommandSlashSubGroups() {
    return _Client.applicationCommandSlashSubGroups;
  }
  get applicationCommandUsers() {
    return _Client.applicationCommandUsers;
  }
  get applicationCommandMessages() {
    return _Client.applicationCommandMessages;
  }
  get applicationCommands() {
    return _Client.applicationCommands;
  }
  get buttonComponents() {
    return _Client.buttonComponents;
  }
  get discords() {
    return _Client.discords;
  }
  get events() {
    return _Client.events;
  }
  get instance() {
    return _Client.instance;
  }
  get modalComponents() {
    return _Client.modalComponents;
  }
  get reactions() {
    return _Client.reactions;
  }
  get selectMenuComponents() {
    return _Client.selectMenuComponents;
  }
  get simpleCommandsByName() {
    return _Client.simpleCommandsByName;
  }
  get simpleCommandMappedPrefix() {
    return _Client.simpleCommandMappedPrefix;
  }
  get simpleCommands() {
    return _Client.simpleCommands;
  }
  // client getters
  get botResolvedGuilds() {
    return resolveIGuilds(this, void 0, this._botGuilds);
  }
  get botGuilds() {
    return this._botGuilds;
  }
  set botGuilds(value) {
    this._botGuilds = value;
  }
  get botId() {
    return this._botId;
  }
  set botId(value) {
    this._botId = value;
  }
  get guards() {
    return this._guards;
  }
  set guards(value) {
    this._guards = value;
  }
  get prefix() {
    return this._prefix;
  }
  set prefix(value) {
    this._prefix = value;
  }
  get simpleCommandConfig() {
    return this._simpleCommandConfig;
  }
  set simpleCommandConfig(value) {
    this._simpleCommandConfig = value;
  }
  get silent() {
    return this._silent;
  }
  set silent(value) {
    this._silent = value;
  }
  /**
   * Extend original client class of discord.js
   *
   * @param options - Client options
   * ___
   *
   * [View Documentation](https://discordx.js.org/docs/discordx/basics/client)
   */
  constructor(options) {
    super(options);
    this._silent = options.silent ?? true;
    this.guards = options.guards ?? [];
    this.botGuilds = options.botGuilds ?? [];
    this._botId = options.botId ?? "bot";
    this._prefix = options.simpleCommand?.prefix ?? ["!"];
    this._simpleCommandConfig = options.simpleCommand;
    this.logger = options.logger ?? console;
  }
  /**
   * Start bot
   *
   * @param token - Bot token
   */
  async login(token) {
    await this.build();
    if (!this.silent) {
      this.logger.log(
        `${this.user?.username ?? this.botId} >> connecting discord...
`
      );
    }
    return super.login(token);
  }
  /**
   * Print information about all events and commands to your console
   */
  printDebug() {
    if (!this.instance.isBuilt) {
      this.logger.error(
        "Build the app before running this method with client.build()"
      );
      return;
    }
    this.logger.log("client >> Events");
    if (this.events.length) {
      this.events.forEach((event) => {
        const eventName = event.event;
        const className = event.classRef.name;
        const key = event.key;
        this.logger.log(`>> ${eventName} (${className}.${key})`);
      });
    } else {
      this.logger.log("	No event detected");
    }
    this.logger.log("");
    this.logger.log("client >> buttons");
    if (this.buttonComponents.length) {
      this.buttonComponents.forEach((btn) => {
        const className = btn.classRef.name;
        const key = btn.key;
        this.logger.log(`>> ${btn.id.toString()} (${className}.${key})`);
      });
    } else {
      this.logger.log("	No buttons detected");
    }
    this.logger.log("");
    this.logger.log("client >> select menu's");
    if (this.selectMenuComponents.length) {
      this.selectMenuComponents.forEach((menu) => {
        const className = menu.classRef.name;
        const key = menu.key;
        this.logger.log(`>> ${menu.id.toString()} (${className}.${key})`);
      });
    } else {
      this.logger.log("	No select menu detected");
    }
    this.logger.log("");
    this.logger.log("client >> modals");
    if (this.modalComponents.length) {
      this.modalComponents.forEach((menu) => {
        const className = menu.classRef.name;
        const key = menu.key;
        this.logger.log(`>> ${menu.id.toString()} (${className}.${key})`);
      });
    } else {
      this.logger.log("	No modal detected");
    }
    this.logger.log("");
    this.logger.log("client >> reactions");
    if (this.reactions.length) {
      this.reactions.forEach((menu) => {
        const className = menu.classRef.name;
        const key = menu.key;
        this.logger.log(`>> ${menu.emoji} (${className}.${key})`);
      });
    } else {
      this.logger.log("	No reaction detected");
    }
    this.logger.log("");
    this.logger.log("client >> context menu's");
    const contexts = [
      ...this.applicationCommandUsers,
      ...this.applicationCommandMessages
    ];
    if (contexts.length) {
      contexts.forEach((menu) => {
        const type = menu.type.toString();
        const className = menu.classRef.name;
        const key = menu.key;
        this.logger.log(`>> ${menu.name} (${type}) (${className}.${key})`);
      });
    } else {
      this.logger.log("	No context menu detected");
    }
    this.logger.log("");
    this.logger.log("client >> application commands");
    if (this.applicationCommands.length) {
      this.applicationCommands.forEach((DCommand, index) => {
        if (DCommand.botIds.length && !DCommand.botIds.includes(this.botId)) {
          return;
        }
        const line = index !== 0 ? "\n" : "";
        const className = DCommand.classRef.name;
        const key = DCommand.key;
        this.logger.log(`${line}	>> ${DCommand.name} (${className}.${key})`);
        const printOptions = (options, depth) => {
          const tab = Array(depth).join("		");
          options.forEach((option, optionIndex) => {
            const className2 = option.classRef.name;
            const key2 = option.key;
            this.logger.log(
              `${(option.type === import_discord2.ApplicationCommandOptionType.Subcommand || option.type === import_discord2.ApplicationCommandOptionType.SubcommandGroup) && optionIndex !== 0 ? "\n" : ""}${tab}>> ${option.type === import_discord2.ApplicationCommandOptionType.Subcommand || option.type === import_discord2.ApplicationCommandOptionType.SubcommandGroup ? option.name : option.name}: ${import_discord2.ApplicationCommandOptionType[option.type].toLowerCase()} (${className2}.${key2})`
            );
            printOptions(option.options, depth + 1);
          });
        };
        printOptions(DCommand.options, 2);
      });
    } else {
      this.logger.log("	No application command detected");
    }
    this.logger.log("");
    this.logger.log("client >> simple commands");
    if (this.simpleCommands.length) {
      this.simpleCommands.forEach((cmd) => {
        const className = cmd.classRef.name;
        const key = cmd.key;
        this.logger.log(`	>> ${cmd.name} (${className}.${key})`);
        if (cmd.aliases.length) {
          this.logger.log(`		aliases:`, cmd.aliases.join(", "));
        }
        const printOptions = (options, depth) => {
          const tab = Array(depth).join("		");
          options.forEach((option) => {
            const type = SimpleCommandOptionType[option.type];
            const className2 = option.classRef.name;
            const key2 = option.key;
            this.logger.log(
              `${tab}${option.name}: ${type} (${className2}.${key2})`
            );
          });
        };
        printOptions(cmd.options, 2);
        this.logger.log("");
      });
    } else {
      this.logger.log("	No simple command detected");
    }
    this.logger.log("\n");
  }
  /**
   * Get commands mapped by guild id (in case of multi bot, commands are filtered for this client only)
   * @returns
   */
  async CommandByGuild() {
    const botResolvedGuilds = await this.botResolvedGuilds;
    const guildDCommandStore = /* @__PURE__ */ new Map();
    const allGuildDCommands = this.applicationCommands.filter((DCommand) => {
      const guilds = [...botResolvedGuilds, ...DCommand.guilds];
      return DCommand.isBotAllowed(this.botId) && guilds.length;
    });
    await Promise.all(
      allGuildDCommands.map(async (DCommand) => {
        const guilds = await resolveIGuilds(this, DCommand, [
          ...botResolvedGuilds,
          ...DCommand.guilds
        ]);
        guilds.forEach((guild) => {
          const commands = guildDCommandStore.get(guild) ?? [];
          return guildDCommandStore.set(guild, [...commands, DCommand]);
        });
      })
    );
    return guildDCommandStore;
  }
  /**
   * Initialize application commands
   */
  async initApplicationCommands(retainDeleted = false) {
    const allGuildPromises = [];
    const guildDCommandStore = await this.CommandByGuild();
    guildDCommandStore.forEach((DCommands, guildId) => {
      const guild = this.guilds.cache.get(guildId);
      if (!guild) {
        return;
      }
      allGuildPromises.push(
        this.initGuildApplicationCommands(guildId, DCommands, retainDeleted)
      );
    });
    await Promise.all([
      Promise.all(allGuildPromises),
      this.initGlobalApplicationCommands(retainDeleted)
    ]);
  }
  /**
   * Init application commands for guild
   */
  async initGuildApplicationCommands(guildId, DCommands, retainDeleted = false) {
    const botResolvedGuilds = await this.botResolvedGuilds;
    const guild = this.guilds.cache.get(guildId);
    if (!guild) {
      this.logger.warn(
        `${this.user?.username ?? this.botId} >> initGuildApplicationCommands: skipped (Reason: guild ${guildId} unavailable)`
      );
      return;
    }
    const discordCommands = await guild.commands.fetch({
      withLocalizations: true
    });
    const commandsToAdd = DCommands.filter((DCommand) => {
      const match = (cmd) => {
        return cmd.name === DCommand.name && cmd.type === DCommand.type;
      };
      return !discordCommands.find(match);
    });
    const commandsToUpdate = [];
    const commandsToSkip = [];
    DCommands.forEach((DCommand) => {
      const match = (cmd) => {
        return cmd.name === DCommand.name && cmd.type === DCommand.type;
      };
      const findCommand = discordCommands.find(match);
      if (!findCommand) {
        return;
      }
      const mixinCommand = new ApplicationCommandMixin(findCommand, DCommand);
      if (!isApplicationCommandEqual(findCommand, DCommand, true)) {
        commandsToUpdate.push(mixinCommand);
      } else {
        commandsToSkip.push(mixinCommand);
      }
    });
    const commandsToDelete = [];
    await Promise.all(
      discordCommands.map(async (cmd) => {
        const match = (DCommand) => {
          return cmd.name === DCommand.name && cmd.type === DCommand.type;
        };
        const DCommandFind = DCommands.find(match);
        if (!DCommandFind) {
          commandsToDelete.push(cmd);
          return;
        }
        const guilds = await resolveIGuilds(this, DCommandFind, [
          ...botResolvedGuilds,
          ...DCommandFind.guilds
        ]);
        if (!cmd.guildId || !guilds.includes(cmd.guildId)) {
          commandsToDelete.push(cmd);
        }
      })
    );
    if (!this.silent) {
      let str = `${this.user?.username ?? "Bot"} >> commands >> guild: #${guild.toString()}`;
      const commandsToAddNames = commandsToAdd.map((DCommand) => DCommand.name).join(", ");
      const commandsToDeleteNames = commandsToDelete.map((DCommand) => DCommand.name).join(", ");
      const commandsToSkipNames = commandsToSkip.map((DCommand) => DCommand.name).join(", ");
      const commandsToUpdateNames = commandsToUpdate.map((DCommand) => DCommand.name).join(", ");
      const deleteOrRetain = retainDeleted ? "retaining" : "deleting";
      str += `
	>> adding   ${String(commandsToAdd.length)} [${commandsToAddNames}]`;
      str += `
	>> ${deleteOrRetain}   ${String(commandsToDelete.length)} [${commandsToDeleteNames}]`;
      str += `
	>> skipping   ${String(commandsToSkip.length)} [${commandsToSkipNames}]`;
      str += `
	>> updating   ${String(commandsToUpdate.length)} [${commandsToUpdateNames}]`;
      str += "\n";
      this.logger.log(str);
    }
    const bulkUpdate = [];
    commandsToSkip.forEach((cmd) => bulkUpdate.push(cmd.instance.toJSON()));
    commandsToAdd.forEach((DCommand) => bulkUpdate.push(DCommand.toJSON()));
    commandsToUpdate.forEach((cmd) => bulkUpdate.push(cmd.instance.toJSON()));
    if (retainDeleted) {
      commandsToDelete.forEach((cmd) => {
        bulkUpdate.push(cmd.toJSON());
      });
    }
    if (bulkUpdate.length === 0) {
      return;
    }
    await guild.commands.set(bulkUpdate);
  }
  /**
   * Init global application commands
   */
  async initGlobalApplicationCommands(retainDeleted = false) {
    const botResolvedGuilds = await this.botResolvedGuilds;
    if (!this.application) {
      throw Error(
        "The client is not yet ready, connect to discord before fetching commands"
      );
    }
    const allDiscordCommands = await this.application.commands.fetch();
    const discordCommands = allDiscordCommands.filter((cmd) => !cmd.guild);
    const DCommands = this.applicationCommands.filter((DCommand) => {
      if (botResolvedGuilds.length || DCommand.guilds.length) {
        return false;
      }
      if (DCommand.botIds.length && !DCommand.botIds.includes(this.botId)) {
        return false;
      }
      return true;
    });
    const commandsToAdd = DCommands.filter((DCommand) => {
      const match = (cmd) => {
        return cmd.name === DCommand.name && cmd.type === DCommand.type;
      };
      return !discordCommands.find(match);
    });
    const commandsToUpdate = [];
    const commandsToSkip = [];
    DCommands.forEach((DCommand) => {
      const match = (cmd) => {
        return cmd.name === DCommand.name && cmd.type === DCommand.type;
      };
      const discordCommand = discordCommands.find(match);
      if (!discordCommand) {
        return;
      }
      const mixinCommand = new ApplicationCommandMixin(
        discordCommand,
        DCommand
      );
      if (!isApplicationCommandEqual(discordCommand, DCommand)) {
        commandsToUpdate.push(mixinCommand);
      } else {
        commandsToSkip.push(mixinCommand);
      }
    });
    const commandsToDelete = discordCommands.filter((cmd) => {
      const match = (DCommand) => {
        return DCommand.name !== cmd.name || DCommand.type !== cmd.type;
      };
      return DCommands.every(match);
    });
    if (!this.silent) {
      let str = `${this.user?.username ?? this.botId} >> commands >> global`;
      const commandsToAddNames = commandsToAdd.map((DCommand) => DCommand.name).join(", ");
      const commandsToDeleteNames = commandsToDelete.map((DCommand) => DCommand.name).join(", ");
      const commandsToSkipNames = commandsToSkip.map((DCommand) => DCommand.name).join(", ");
      const commandsToUpdateNames = commandsToUpdate.map((DCommand) => DCommand.name).join(", ");
      const deleteOrRetain = retainDeleted ? "retaining" : "deleting";
      str += `
	>> adding   ${String(commandsToAdd.length)} [${commandsToAddNames}]`;
      str += `
	>> ${deleteOrRetain}   ${String(commandsToDelete.size)} [${commandsToDeleteNames}]`;
      str += `
	>> skipping   ${String(commandsToSkip.length)} [${commandsToSkipNames}]`;
      str += `
	>> updating   ${String(commandsToUpdate.length)} [${commandsToUpdateNames}]`;
      str += "\n";
      this.logger.log(str);
    }
    const bulkUpdate = [];
    commandsToSkip.forEach((cmd) => bulkUpdate.push(cmd.instance.toJSON()));
    commandsToAdd.forEach((instance) => bulkUpdate.push(instance.toJSON()));
    commandsToUpdate.forEach((cmd) => bulkUpdate.push(cmd.instance.toJSON()));
    if (retainDeleted) {
      commandsToDelete.forEach((cmd) => {
        bulkUpdate.push(cmd.toJSON());
      });
    }
    if (bulkUpdate.length === 0) {
      return;
    }
    await this.application.commands.set(bulkUpdate);
  }
  /**
   * Clear the application commands globally or for some guilds
   *
   * @param guilds - The guild Ids (empty -> globally)
   */
  async clearApplicationCommands(...guilds) {
    if (guilds.length) {
      await Promise.all(
        // Select and delete the commands of each guild
        guilds.map((guild) => this.guilds.cache.get(guild)?.commands.set([]))
      );
    } else {
      await this.application?.commands.set([]);
    }
  }
  /**
   * Get the group tree of an slash interaction
   * /hello => ["hello"]
   * /test hello => ["test", "hello"]
   * /test hello me => ["test", "hello", "me"]
   *
   * @param interaction - The targeted slash interaction
   *
   * @returns
   */
  getApplicationCommandGroupTree(interaction) {
    const tree = [];
    const getOptionsTree = (option) => {
      if (!option) {
        return;
      }
      if (!option.type || option.type === import_discord2.ApplicationCommandOptionType.SubcommandGroup || option.type === import_discord2.ApplicationCommandOptionType.Subcommand) {
        if (option.name) {
          tree.push(option.name);
        }
        getOptionsTree(Array.from(option.options?.values() ?? [])[0]);
      }
    };
    getOptionsTree({
      name: interaction.commandName,
      options: Array.from(interaction.options.data.values()),
      type: void 0
    });
    return tree;
  }
  /**
   * Return the corresponding @Slash from a tree
   *
   * @param tree - Array of string
   *
   * @returns
   */
  getApplicationCommandFromTree(tree) {
    return this.applicationCommandSlashesFlat.find((slash) => {
      switch (tree.length) {
        case 1:
          return slash.group === void 0 && slash.subgroup === void 0 && slash.name === tree[0] && slash.type === import_discord2.ApplicationCommandType.ChatInput;
        case 2:
          return slash.group === tree[0] && slash.subgroup === void 0 && slash.name === tree[1] && slash.type === import_discord2.ApplicationCommandType.ChatInput;
        case 3:
          return slash.group === tree[0] && slash.subgroup === tree[1] && slash.name === tree[2] && slash.type === import_discord2.ApplicationCommandType.ChatInput;
        default:
          return false;
      }
    });
  }
  /**
   * Execute all types of interaction
   *
   * @param interaction - Interaction
   *
   * @returns
   */
  executeInteraction(interaction) {
    if (interaction.isButton()) {
      return this.executeComponent(this.buttonComponents, interaction);
    }
    if (interaction.type === import_discord2.InteractionType.ModalSubmit) {
      return this.executeComponent(this.modalComponents, interaction);
    }
    if (interaction.isAnySelectMenu()) {
      return this.executeComponent(this.selectMenuComponents, interaction);
    }
    if (interaction.isContextMenuCommand()) {
      return this.executeContextMenu(interaction);
    }
    return this.executeCommandInteraction(interaction);
  }
  /**
   * Execute command interaction
   *
   * @param interaction - Interaction instance
   *
   * @returns
   */
  async executeCommandInteraction(interaction) {
    const tree = this.getApplicationCommandGroupTree(interaction);
    const applicationCommand = this.getApplicationCommandFromTree(tree);
    if (!applicationCommand?.isBotAllowed(this.botId)) {
      if (!this.silent) {
        this.logger.warn(
          `${this.user?.username ?? this.botId} >> interaction not found, commandName: ${interaction.commandName}`
        );
      }
      return null;
    }
    if (interaction.type === import_discord2.InteractionType.ApplicationCommandAutocomplete) {
      const focusOption = interaction.options.getFocused(true);
      const option = applicationCommand.options.find(
        (op) => op.name === focusOption.name
      );
      if (option && typeof option.autocomplete === "function") {
        await option.autocomplete.call(
          import_di.DIService.engine.getService(option.from),
          interaction,
          applicationCommand
        );
        return null;
      }
    }
    return applicationCommand.execute(this.guards, interaction, this);
  }
  /**
   * Execute component interaction
   *
   * @param interaction - Interaction instance
   *
   * @returns
   */
  async executeComponent(components, interaction) {
    const executes = components.filter((component) => {
      return component.isId(interaction.customId) && component.isBotAllowed(this.botId);
    });
    if (!executes.length) {
      if (!this.silent) {
        this.logger.warn(
          `${this.user?.username ?? this.botId} >> ${interaction.isButton() ? "button" : interaction.isAnySelectMenu() ? "select menu" : "modal"} component handler not found, interactionId: ${interaction.id} | customId: ${interaction.customId}`
        );
      }
      return null;
    }
    const results = await Promise.all(
      executes.map(async (component) => {
        if (!await component.isGuildAllowed(this, interaction.guildId)) {
          return null;
        }
        return component.execute(this.guards, interaction, this);
      })
    );
    return results;
  }
  /**
   * Execute context menu interaction
   *
   * @param interaction - Interaction instance
   *
   * @returns
   */
  executeContextMenu(interaction) {
    const applicationCommand = interaction.isUserContextMenuCommand() ? this.applicationCommandUsers.find(
      (cmd) => cmd.name === interaction.commandName
    ) : this.applicationCommandMessages.find(
      (cmd) => cmd.name === interaction.commandName
    );
    if (!applicationCommand?.isBotAllowed(this.botId)) {
      if (!this.silent) {
        this.logger.warn(
          `${this.user?.username ?? this.botId} >> context interaction not found, name: ${interaction.commandName}`
        );
      }
      return null;
    }
    return applicationCommand.execute(this.guards, interaction, this);
  }
  /**
   * Fetch prefix for message
   *
   * @param message - Message instance
   *
   * @returns
   */
  async getMessagePrefix(message) {
    if (typeof this.prefix !== "function") {
      return toStringArray(this.prefix);
    }
    const prefix = await this.prefix(message);
    return toStringArray(prefix);
  }
  /**
   * Parse command message
   *
   * @param prefix - Command prefix
   * @param message - Original message
   * @param caseSensitive - Execute case-sensitively
   *
   * @returns
   */
  async parseCommand(message, caseSensitive = false) {
    const prefix = await this.getMessagePrefix(message);
    const prefixRegex = RegExp(
      `^(${toStringArray(prefix, Array.from(this.simpleCommandMappedPrefix)).map((pfx) => (0, import_escapeRegExp.default)(pfx)).join("|")})`
    );
    const isCommand = prefixRegex.test(message.content);
    if (!isCommand) {
      return 0 /* notCommand */;
    }
    const matchedPrefix = prefixRegex.exec(message.content)?.at(1) ?? "unknown";
    const contentWithoutPrefix = `${message.content.replace(prefixRegex, "").trim()} `;
    const commandRaw = this.simpleCommandsByName.find((cmd) => {
      if (caseSensitive) {
        return contentWithoutPrefix.startsWith(`${cmd.name} `);
      }
      return contentWithoutPrefix.toLowerCase().startsWith(`${cmd.name.toLowerCase()} `);
    });
    if (!commandRaw) {
      return 1 /* notFound */;
    }
    const commandArgs = contentWithoutPrefix.replace(new RegExp(commandRaw.name, "i"), "").trim();
    const command = new SimpleCommandMessage(
      matchedPrefix,
      commandArgs,
      message,
      commandRaw.command,
      this.simpleCommandConfig?.argSplitter
    );
    command.options = await command.resolveOptions();
    return command;
  }
  /**
   * Execute the corresponding @SimpleCommand based on an message instance
   *
   * @param message - Message instance
   * @param options - Options
   *
   * @returns
   */
  async executeCommand(message, caseSensitive) {
    const command = await this.parseCommand(message, caseSensitive ?? false);
    if (command === 0 /* notCommand */) {
      return null;
    }
    if (command === 1 /* notFound */) {
      const handleNotFound = this.simpleCommandConfig?.responses?.notFound;
      if (handleNotFound) {
        if (typeof handleNotFound === "string") {
          await message.reply(handleNotFound);
        } else {
          await handleNotFound(message);
        }
      }
      return null;
    }
    if (!command.info.isBotAllowed(this.botId)) {
      return null;
    }
    if (!await command.info.isGuildAllowed(this, command, message.guildId)) {
      return null;
    }
    if (!command.info.directMessage && !message.guild) {
      return null;
    }
    return command.info.execute(this.guards, command, this);
  }
  /**
   * Parse reaction
   *
   * @param message - Original reaction
   *
   * @returns
   */
  parseReaction(message) {
    const reaction = this.reactions.find((react) => {
      const validNames = [react.emoji, ...react.aliases];
      const { emoji } = message;
      return (emoji.id ? validNames.includes(emoji.id) : false) || (emoji.name ? validNames.includes(emoji.name) : false);
    });
    return reaction;
  }
  /**
   * Execute the corresponding @Reaction based on an message reaction instance
   *
   * @param reaction - MessageReaction instance
   * @param options - Options
   *
   * @returns
   */
  async executeReaction(reaction, user) {
    const action = this.parseReaction(reaction);
    if (!action) {
      return null;
    }
    if (!action.isBotAllowed(this.botId)) {
      return null;
    }
    if (!await action.isGuildAllowed(this, reaction.message.guildId)) {
      return null;
    }
    if (!action.directMessage && !reaction.message.guild) {
      return null;
    }
    if (!action.partial && reaction.partial) {
      reaction = await reaction.fetch();
    }
    if (!action.partial && user.partial) {
      user = await user.fetch();
    }
    if (action.remove) {
      await reaction.users.remove(user.id);
    }
    return action.execute(this.guards, reaction, user, this);
  }
  /**
   * Trigger an event manually (used for testing)
   *
   * @param options - Event data
   * @param params - Params to inject
   *
   * @returns
   */
  trigger(options, params) {
    return this.instance.trigger(options)(params);
  }
  /**
   * Bind discordx events to client
   */
  initEvents() {
    for (const { event, once, rest } of this.instance.usedEvents) {
      const trigger = this.instance.trigger({
        client: this,
        event,
        guards: this.guards,
        once,
        rest
      });
      if (!this._listeners.has(event)) {
        this._listeners.set(event, []);
      }
      this._listeners.get(event)?.push({ once, rest, trigger });
      const method = once ? "once" : "on";
      if (rest) {
        this.rest[method](event, trigger);
      } else {
        this[method](event, trigger);
      }
    }
  }
  /**
   * Unbind all discordx events initialized by the initEvents method.
   */
  removeEvents() {
    this._listeners.forEach((listenerDetails, event) => {
      listenerDetails.forEach(({ rest, trigger }) => {
        if (rest) {
          this.rest.off(event, trigger);
        } else {
          this.off(event, trigger);
        }
      });
    });
    this._listeners.clear();
  }
  /**
   * Manually build client
   */
  async build() {
    if (this._isBuilt) {
      return;
    }
    this._isBuilt = true;
    await this.instance.build();
    this.initEvents();
    if (!this.silent) {
      this.printDebug();
    }
  }
};

// src/decorators/classes/DApplicationCommand.ts
var import_discord3 = require("discord.js");
var import_discord4 = require("discord.js");

// src/decorators/classes/Method.ts
var import_internal = require("@discordx/internal");
var Method = class extends import_internal.Decorator {
  _discord;
  _guards = [];
  get discord() {
    return this._discord;
  }
  set discord(value) {
    this._discord = value;
  }
  /**
   * Compiled methods executes all the guards and the main method
   * ```ts
   * compiledMethod = async (params: ArgsOf<any>, client: Client) => {
   *   guard1(params, client)
   *   guard2(params, client)
   *   guard3(params, client)
   *   main(params, client)
   * }
   * ```
   * @returns
   */
  get execute() {
    return (guards, ...params) => {
      const globalGuards = guards.map(
        (guard) => DGuard.create(guard.bind(void 0))
      );
      return this.getGuardFunction(globalGuards)(...params);
    };
  }
  /**
   * Returns all the guards of the application
   * The guards that are defined globally with Client
   * The guards that decorate @Discord
   * The guards that decorate the method (this)
   */
  get guards() {
    return [
      ...this.discord.guards,
      ...this._guards,
      DGuard.create(this._method?.bind(this._discord.instance))
    ];
  }
  set guards(value) {
    this._guards = value;
  }
  /**
   * Execute a guard with params
   */
  getGuardFunction(globalGuards) {
    const next = async (params, index, paramsToNext) => {
      const nextFn = () => next(params, index + 1, paramsToNext);
      const guardToExecute = [...globalGuards, ...this.guards][index];
      let res;
      if (index >= [...globalGuards, ...this.guards].length - 1) {
        const parsedParams = await this.parseParams(...params);
        res = await (guardToExecute?.fn)(
          ...parsedParams,
          ...params,
          paramsToNext
        );
      } else {
        res = await (guardToExecute?.fn)(
          ...params,
          nextFn,
          paramsToNext
        );
      }
      if (res) {
        return res;
      }
      return paramsToNext;
    };
    return (...params) => next(params, 0, {});
  }
};

// src/decorators/classes/DApplicationCommand.ts
var DApplicationCommand = class _DApplicationCommand extends Method {
  _botIds;
  _contexts;
  _defaultMemberPermissions;
  _description;
  _descriptionLocalizations;
  _dmPermission;
  _group;
  _guilds;
  _integrationTypes;
  _name;
  _nameLocalizations;
  _nsfw;
  _options = [];
  _subgroup;
  _type;
  get botIds() {
    return this._botIds;
  }
  set botIds(value) {
    this._botIds = value;
  }
  get description() {
    return this._description;
  }
  set description(value) {
    this._description = value;
  }
  get defaultMemberPermissions() {
    return this._defaultMemberPermissions;
  }
  set defaultMemberPermissions(value) {
    this._defaultMemberPermissions = value;
  }
  get dmPermission() {
    return this._dmPermission;
  }
  set dmPermission(value) {
    this._dmPermission = value;
  }
  get contexts() {
    return this._contexts;
  }
  set contexts(value) {
    this._contexts = value;
  }
  get integrationTypes() {
    return this._integrationTypes;
  }
  set integrationTypes(value) {
    this._integrationTypes = value;
  }
  get descriptionLocalizations() {
    return this._descriptionLocalizations;
  }
  set descriptionLocalizations(value) {
    this._descriptionLocalizations = value;
  }
  get group() {
    return this._group;
  }
  set group(value) {
    this._group = value;
  }
  get guilds() {
    return this._guilds;
  }
  set guilds(value) {
    this._guilds = value;
  }
  get name() {
    return this._name;
  }
  set name(value) {
    this._name = value;
  }
  get nameLocalizations() {
    return this._nameLocalizations;
  }
  set nameLocalizations(value) {
    this._nameLocalizations = value;
  }
  get nsfw() {
    return this._nsfw;
  }
  set nsfw(value) {
    this._nsfw = value;
  }
  get options() {
    return this._options;
  }
  set options(value) {
    this._options = value;
  }
  get subgroup() {
    return this._subgroup;
  }
  set subgroup(value) {
    this._subgroup = value;
  }
  get type() {
    return this._type;
  }
  set type(value) {
    this._type = value;
  }
  constructor(data) {
    super();
    this._botIds = data.botIds ?? [];
    this._contexts = data.contexts ?? null;
    this._defaultMemberPermissions = data.defaultMemberPermissions ?? null;
    this._description = data.description;
    this._descriptionLocalizations = data.descriptionLocalizations ?? null;
    this._dmPermission = data.dmPermission ?? true;
    this._guilds = data.guilds ?? [];
    this._integrationTypes = data.integrationTypes ?? [
      import_discord3.ApplicationIntegrationType.GuildInstall
    ];
    this._name = data.name;
    this._nameLocalizations = data.nameLocalizations ?? null;
    this._nsfw = data.nsfw ?? false;
    this._type = data.type;
  }
  static create(data) {
    return new _DApplicationCommand(data);
  }
  isBotAllowed(botId) {
    if (!this.botIds.length) {
      return true;
    }
    return this.botIds.includes(botId);
  }
  async getGuilds(client) {
    const guilds = await resolveIGuilds(client, this, [
      ...client.botGuilds,
      ...this.guilds
    ]);
    return guilds;
  }
  async isGuildAllowed(client, guildId) {
    if (!guildId) {
      return true;
    }
    const guilds = await this.getGuilds(client);
    if (!guilds.length) {
      return true;
    }
    return guilds.includes(guildId);
  }
  toSubCommand() {
    const option = DApplicationCommandOption.create({
      description: this.description,
      descriptionLocalizations: this.descriptionLocalizations,
      name: this.name,
      nameLocalizations: this.nameLocalizations,
      type: import_discord4.ApplicationCommandOptionType.Subcommand
    }).decorate(this.classRef, this.key, this.method, this.from, this.index);
    option.options = this.options;
    return option;
  }
  toJSON() {
    const options = [...this.options].reverse().sort((a, b) => {
      if ((a.type === import_discord4.ApplicationCommandOptionType.Subcommand || a.type === import_discord4.ApplicationCommandOptionType.SubcommandGroup) && (b.type === import_discord4.ApplicationCommandOptionType.Subcommand || b.type === import_discord4.ApplicationCommandOptionType.SubcommandGroup)) {
        return a.name < b.name ? -1 : 1;
      }
      return 0;
    }).map((option) => option.toJSON());
    const data = {
      contexts: this.contexts,
      defaultMemberPermissions: this.defaultMemberPermissions,
      description: this.description,
      descriptionLocalizations: this.descriptionLocalizations,
      dmPermission: this.dmPermission,
      integrationTypes: this.integrationTypes,
      name: this.name,
      nameLocalizations: this.nameLocalizations,
      nsfw: this.nsfw,
      options,
      type: this.type
    };
    return data;
  }
  parseParams(interaction) {
    return Promise.all(
      [...this.options].reverse().map((op) => op.parse(interaction))
    );
  }
};

// src/decorators/classes/DApplicationCommandGroup.ts
var import_internal2 = require("@discordx/internal");
var DApplicationCommandGroup = class _DApplicationCommandGroup extends import_internal2.Decorator {
  name;
  root;
  payload;
  constructor(options) {
    super();
    this.name = options.name;
    this.root = options.root;
    this.payload = options.payload;
  }
  static create(options) {
    return new _DApplicationCommandGroup(options);
  }
};

// src/decorators/classes/DApplicationCommandOption.ts
var import_internal3 = require("@discordx/internal");
var import_discord5 = require("discord.js");
var DApplicationCommandOption = class _DApplicationCommandOption extends import_internal3.Decorator {
  _autocomplete;
  _channelTypes = void 0;
  _choices = [];
  _description;
  _descriptionLocalizations;
  _name;
  _nameLocalizations;
  _maxValue;
  _minValue;
  _maxLength;
  _minLength;
  _options = [];
  _required = true;
  _type;
  _transformer;
  get autocomplete() {
    return this._autocomplete;
  }
  set autocomplete(value) {
    this._autocomplete = value;
  }
  get channelTypes() {
    return this._channelTypes;
  }
  set channelTypes(value) {
    this._channelTypes = value;
  }
  get choices() {
    return this._choices;
  }
  set choices(value) {
    this._choices = value;
  }
  get description() {
    return this._description;
  }
  set description(value) {
    this._description = value;
  }
  get descriptionLocalizations() {
    return this._descriptionLocalizations;
  }
  set descriptionLocalizations(value) {
    this._descriptionLocalizations = value;
  }
  get isNode() {
    return this.type === import_discord5.ApplicationCommandOptionType.Subcommand || this.type === import_discord5.ApplicationCommandOptionType.SubcommandGroup;
  }
  get maxValue() {
    return this._maxValue;
  }
  set maxValue(value) {
    this._maxValue = value;
  }
  get minValue() {
    return this._minValue;
  }
  set minValue(value) {
    this._minValue = value;
  }
  get maxLength() {
    return this._maxLength;
  }
  set maxLength(value) {
    this._maxLength = value;
  }
  get minLength() {
    return this._minLength;
  }
  set minLength(value) {
    this._minLength = value;
  }
  get name() {
    return this._name;
  }
  set name(value) {
    this._name = value;
  }
  get nameLocalizations() {
    return this._nameLocalizations;
  }
  set nameLocalizations(value) {
    this._nameLocalizations = value;
  }
  get options() {
    return this._options;
  }
  set options(value) {
    this._options = value;
  }
  get required() {
    return this._required;
  }
  set required(value) {
    this._required = value;
  }
  get type() {
    return this._type;
  }
  set type(value) {
    this._type = value;
  }
  constructor(data) {
    super();
    this._name = data.name;
    this._autocomplete = data.autocomplete;
    this._channelTypes = data.channelType?.sort();
    this._choices = data.choices ?? [];
    this._description = data.description;
    this._index = data.index;
    this._maxValue = data.maxValue;
    this._minValue = data.minValue;
    this._maxLength = data.maxLength;
    this._minLength = data.minLength;
    this._required = data.required ?? false;
    this._type = data.type;
    this._descriptionLocalizations = data.descriptionLocalizations ?? null;
    this._nameLocalizations = data.nameLocalizations ?? null;
    this._transformer = data.transformer;
  }
  static create(data) {
    return new _DApplicationCommandOption(data);
  }
  toJSON() {
    const options = [...this.options].reverse().map((option) => option.toJSON());
    const data = {
      autocomplete: this.autocomplete ? true : void 0,
      channelTypes: this.channelTypes,
      choices: this.isNode ? void 0 : this.choices.length === 0 ? void 0 : this.choices.map((choice) => choice.toJSON()),
      description: this.description,
      descriptionLocalizations: this.descriptionLocalizations,
      maxLength: this.maxLength,
      maxValue: this.maxValue,
      minLength: this.minLength,
      minValue: this.minValue,
      name: this.name,
      nameLocalizations: this.nameLocalizations,
      options: options.length === 0 ? void 0 : options,
      required: this.isNode ? void 0 : this.required,
      type: this.type
    };
    return data;
  }
  parseType(interaction) {
    switch (this.type) {
      case import_discord5.ApplicationCommandOptionType.Attachment:
        return interaction.options.getAttachment(this.name) ?? void 0;
      case import_discord5.ApplicationCommandOptionType.String:
        return interaction.options.getString(this.name) ?? void 0;
      case import_discord5.ApplicationCommandOptionType.Boolean:
        return interaction.options.getBoolean(this.name) ?? void 0;
      case import_discord5.ApplicationCommandOptionType.Number:
        return interaction.options.getNumber(this.name) ?? void 0;
      case import_discord5.ApplicationCommandOptionType.Integer:
        return interaction.options.getInteger(this.name) ?? void 0;
      case import_discord5.ApplicationCommandOptionType.Role:
        return interaction.options.getRole(this.name) ?? void 0;
      case import_discord5.ApplicationCommandOptionType.Channel:
        return interaction.options.getChannel(this.name) ?? void 0;
      case import_discord5.ApplicationCommandOptionType.Mentionable:
        return interaction.options.getMentionable(this.name) ?? void 0;
      case import_discord5.ApplicationCommandOptionType.User:
        return interaction.options.getMember(this.name) ?? interaction.options.getUser(this.name) ?? void 0;
      default:
        return interaction.options.getString(this.name) ?? void 0;
    }
  }
  parse(interaction) {
    if (this._transformer !== void 0) {
      return this._transformer(this.parseType(interaction), interaction);
    }
    return this.parseType(interaction);
  }
};

// src/decorators/classes/DApplicationCommandOptionChoice.ts
var import_internal4 = require("@discordx/internal");
var DApplicationCommandOptionChoice = class _DApplicationCommandOptionChoice extends import_internal4.Decorator {
  _name;
  _nameLocalizations;
  _value;
  get name() {
    return this._name;
  }
  set name(value) {
    this._name = value;
  }
  get nameLocalizations() {
    return this._nameLocalizations;
  }
  set nameLocalizations(value) {
    this._nameLocalizations = value;
  }
  get value() {
    return this._value;
  }
  set value(value) {
    this._value = value;
  }
  constructor(data) {
    super();
    this._name = data.name;
    this._nameLocalizations = data.nameLocalizations ?? null;
    this._value = data.value ?? data.name;
  }
  static create(data) {
    return new _DApplicationCommandOptionChoice(data);
  }
  toJSON() {
    return {
      name: this.name,
      nameLocalizations: this.nameLocalizations,
      value: this.value
    };
  }
};

// src/decorators/classes/DComponent.ts
var DComponent = class _DComponent extends Method {
  _type;
  _id;
  _guilds;
  _botIds;
  get type() {
    return this._type;
  }
  get botIds() {
    return this._botIds;
  }
  set botIds(value) {
    this._botIds = value;
  }
  get id() {
    return this._id;
  }
  set id(value) {
    this._id = value;
  }
  get guilds() {
    return this._guilds;
  }
  set guilds(value) {
    this._guilds = value;
  }
  constructor(data) {
    super();
    this._type = data.type;
    this._id = data.id;
    this._guilds = data.guilds ?? [];
    this._botIds = data.botIds ?? [];
  }
  static create(data) {
    return new _DComponent(data);
  }
  isBotAllowed(botId) {
    if (!this.botIds.length) {
      return true;
    }
    return this.botIds.includes(botId);
  }
  async getGuilds(client) {
    const guilds = await resolveIGuilds(client, this, [
      ...client.botGuilds,
      ...this.guilds
    ]);
    return guilds;
  }
  async isGuildAllowed(client, guildId) {
    if (!guildId) {
      return true;
    }
    const guilds = await this.getGuilds(client);
    if (!guilds.length) {
      return true;
    }
    return guilds.includes(guildId);
  }
  isId(text) {
    return typeof this.id === "string" ? this.id === text : this.id.test(text);
  }
  parseParams() {
    return [];
  }
};

// src/decorators/classes/DDiscord.ts
var import_di2 = require("@discordx/di");
var import_internal5 = require("@discordx/internal");
var DDiscord = class _DDiscord extends import_internal5.Decorator {
  _applicationCommands = [];
  _botIds = [];
  _buttonComponents = [];
  _description;
  _events = [];
  _guards = [];
  _guilds = [];
  _modalComponents = [];
  _name;
  _reactions = [];
  _selectMenuComponents = [];
  _simpleCommands = [];
  get applicationCommands() {
    return this._applicationCommands;
  }
  set applicationCommands(value) {
    this._applicationCommands = value;
  }
  get botIds() {
    return this._botIds;
  }
  set botIds(value) {
    this._botIds = value;
  }
  get buttons() {
    return this._buttonComponents;
  }
  set buttons(value) {
    this._buttonComponents = value;
  }
  get description() {
    return this._description;
  }
  set description(value) {
    this._description = value;
  }
  get events() {
    return this._events;
  }
  set events(value) {
    this._events = value;
  }
  get guards() {
    return this._guards;
  }
  set guards(value) {
    this._guards = value;
  }
  get guilds() {
    return this._guilds;
  }
  set guilds(value) {
    this._guilds = value;
  }
  get instance() {
    return import_di2.DIService.engine.getService(this.from);
  }
  get modal() {
    return this._modalComponents;
  }
  set modal(value) {
    this._modalComponents = value;
  }
  get name() {
    return this._name;
  }
  set name(value) {
    this._name = value;
  }
  get reactions() {
    return this._reactions;
  }
  set reactions(value) {
    this._reactions = value;
  }
  get selectMenus() {
    return this._selectMenuComponents;
  }
  set selectMenus(value) {
    this._selectMenuComponents = value;
  }
  get simpleCommands() {
    return this._simpleCommands;
  }
  set simpleCommands(value) {
    this._simpleCommands = value;
  }
  constructor(name, description) {
    super();
    this._name = name;
    this._description = description ?? name;
  }
  static create(name, description) {
    return new _DDiscord(name, description);
  }
};

// src/decorators/classes/DGuard.ts
var import_internal6 = require("@discordx/internal");
var DGuard = class _DGuard extends import_internal6.Decorator {
  _fn;
  get fn() {
    return this._fn;
  }
  constructor(fn) {
    super();
    this._fn = fn;
  }
  static create(fn) {
    return new _DGuard(fn);
  }
};

// src/decorators/classes/DOn.ts
var DOn = class _DOn extends Method {
  _event;
  _once;
  _rest;
  _priority;
  _botIds;
  get botIds() {
    return this._botIds;
  }
  set botIds(value) {
    this._botIds = value;
  }
  get event() {
    return this._event;
  }
  set event(value) {
    this._event = value;
  }
  get once() {
    return this._once;
  }
  set once(value) {
    this._once = value;
  }
  get priority() {
    return this._priority;
  }
  set priority(value) {
    this._priority = value;
  }
  get rest() {
    return this._rest;
  }
  set rest(value) {
    this._rest = value;
  }
  constructor(data) {
    super();
    this._event = data.event;
    this._once = data.once;
    this._priority = data.priority ?? Number.MAX_SAFE_INTEGER;
    this._rest = data.rest;
    this._botIds = data.botIds ?? [];
  }
  static create(data) {
    return new _DOn(data);
  }
  isBotAllowed(botId) {
    if (!this.botIds.length) {
      return true;
    }
    return this.botIds.includes(botId);
  }
  parseParams() {
    return [];
  }
};

// src/decorators/classes/DReaction.ts
var DReaction = class _DReaction extends Method {
  _emoji;
  _description;
  _directMessage;
  _guilds;
  _botIds;
  _aliases;
  _remove;
  _partial;
  get aliases() {
    return this._aliases;
  }
  set aliases(value) {
    this._aliases = value;
  }
  get botIds() {
    return this._botIds;
  }
  set botIds(value) {
    this._botIds = value;
  }
  get guilds() {
    return this._guilds;
  }
  set guilds(value) {
    this._guilds = value;
  }
  get directMessage() {
    return this._directMessage;
  }
  set directMessage(value) {
    this._directMessage = value;
  }
  get emoji() {
    return this._emoji;
  }
  set emoji(value) {
    this._emoji = value;
  }
  get description() {
    return this._description;
  }
  set description(value) {
    this._description = value;
  }
  get remove() {
    return this._remove;
  }
  set remove(value) {
    this._remove = value;
  }
  get partial() {
    return this._partial;
  }
  set partial(value) {
    this._partial = value;
  }
  constructor(data) {
    super();
    this._emoji = data.emoji;
    this._description = data.description ?? this.emoji;
    this._directMessage = data.directMessage ?? true;
    this._guilds = data.guilds ?? [];
    this._botIds = data.botIds ?? [];
    this._aliases = data.aliases ?? [];
    this._remove = data.remove ?? false;
    this._partial = data.partial ?? false;
  }
  static create(data) {
    return new _DReaction(data);
  }
  isBotAllowed(botId) {
    if (!this.botIds.length) {
      return true;
    }
    return this.botIds.includes(botId);
  }
  async getGuilds(client) {
    const guilds = await resolveIGuilds(client, this, [
      ...client.botGuilds,
      ...this.guilds
    ]);
    return guilds;
  }
  async isGuildAllowed(client, guildId) {
    if (!guildId) {
      return true;
    }
    const guilds = await this.getGuilds(client);
    if (!guilds.length) {
      return true;
    }
    return guilds.includes(guildId);
  }
  parseParams() {
    return [];
  }
};

// src/decorators/classes/DSimpleCommand.ts
var import_discord6 = require("discord.js");
var DSimpleCommand = class _DSimpleCommand extends Method {
  _description;
  _name;
  _prefix;
  _directMessage;
  _argSplitter;
  _options = [];
  _guilds;
  _botIds;
  _aliases;
  get aliases() {
    return this._aliases;
  }
  set aliases(value) {
    this._aliases = value;
  }
  get botIds() {
    return this._botIds;
  }
  set botIds(value) {
    this._botIds = value;
  }
  get prefix() {
    return this._prefix;
  }
  set prefix(value) {
    this._prefix = value;
  }
  get guilds() {
    return this._guilds;
  }
  set guilds(value) {
    this._guilds = value;
  }
  get argSplitter() {
    return this._argSplitter;
  }
  set argSplitter(value) {
    this._argSplitter = value;
  }
  get directMessage() {
    return this._directMessage;
  }
  set directMessage(value) {
    this._directMessage = value;
  }
  get name() {
    return this._name;
  }
  set name(value) {
    this._name = value;
  }
  get description() {
    return this._description;
  }
  set description(value) {
    this._description = value;
  }
  get options() {
    return this._options;
  }
  set options(value) {
    this._options = value;
  }
  constructor(data) {
    super();
    this._name = data.name;
    this._description = data.description ?? this.name;
    this._directMessage = data.directMessage ?? true;
    this._argSplitter = data.argSplitter;
    this._options = [];
    this._prefix = data.prefix;
    this._guilds = data.guilds ?? [];
    this._botIds = data.botIds ?? [];
    this._aliases = data.aliases ?? [];
  }
  static create(data) {
    return new _DSimpleCommand(data);
  }
  isBotAllowed(botId) {
    if (!this.botIds.length) {
      return true;
    }
    return this.botIds.includes(botId);
  }
  async getGuilds(client, command) {
    const guilds = await resolveIGuilds(client, command, [
      ...client.botGuilds,
      ...this.guilds
    ]);
    return guilds;
  }
  async isGuildAllowed(client, command, guildId) {
    if (!guildId) {
      return true;
    }
    const guilds = await this.getGuilds(client, command);
    if (!guilds.length) {
      return true;
    }
    return guilds.includes(guildId);
  }
  parseParams(command) {
    return command.options;
  }
  parseParamsEx(command) {
    if (!this.options.length) {
      return Promise.resolve([]);
    }
    const splitterEx = this.argSplitter ?? command.splitter ?? " ";
    const args = typeof splitterEx === "function" ? splitterEx(command) : command.argString.split(splitterEx).filter((op) => op.length).map((op) => op.trim());
    return Promise.all(
      this.options.sort((a, b) => (a.index ?? 0) - (b.index ?? 0)).map(async (op, index) => {
        const option = args[index];
        const id = option?.replace(/\D/g, "");
        const validId = id !== void 0 && id.length >= 16 && id.length <= 20;
        if (!option?.length) {
          return null;
        }
        if (op.type === 2 /* Boolean */) {
          if (option.toLocaleLowerCase() === "false" || option.toLocaleLowerCase() === "0") {
            return false;
          }
          return Boolean(option);
        }
        if (op.type === 1 /* Number */) {
          return Number(option);
        }
        if (op.type === 4 /* Channel */) {
          if (!validId || !command.message.guild) {
            return null;
          }
          return command.message.guild.channels.fetch(id).catch(() => null);
        }
        if (op.type === 5 /* Role */) {
          if (!validId || !command.message.guild) {
            return null;
          }
          return command.message.guild.roles.fetch(id).catch(() => null);
        }
        if (op.type === 3 /* User */) {
          if (!validId) {
            return null;
          }
          if (command.message.channel.type === import_discord6.ChannelType.DM) {
            if (command.message.client.user.id === id) {
              return command.message.client.user;
            }
            if (command.message.author.id === id) {
              return command.message.author;
            }
            return null;
          }
          if (!command.message.guild) {
            return null;
          }
          return command.message.guild.members.fetch(id).catch(() => null);
        }
        if (op.type === 6 /* Mentionable */) {
          if (!validId) {
            return null;
          }
          if (command.message.channel.type === import_discord6.ChannelType.DM) {
            if (command.message.client.user.id === id) {
              return command.message.client.user;
            }
            if (command.message.author.id === id) {
              return command.message.author;
            }
            return null;
          }
          if (!command.message.guild) {
            return null;
          }
          const member = await command.message.guild.members.fetch(id).catch(() => null);
          if (member) {
            return member;
          }
          const role = await command.message.guild.roles.fetch(id).catch(() => null);
          return role;
        }
        return option;
      })
    );
  }
};

// src/decorators/classes/DSimpleCommandOption.ts
var import_internal7 = require("@discordx/internal");
var DSimpleCommandOption = class _DSimpleCommandOption extends import_internal7.Decorator {
  _name;
  _description;
  _type;
  get name() {
    return this._name;
  }
  set name(value) {
    this._name = value;
  }
  get type() {
    return this._type;
  }
  set type(value) {
    this._type = value;
  }
  get description() {
    return this._description;
  }
  set description(value) {
    this._description = value;
  }
  constructor(data) {
    super();
    this._name = data.name;
    this._type = data.type ?? 0 /* String */;
    this._description = data.description ?? SimpleCommandOptionType[this._type].toLowerCase();
  }
  static create(data) {
    return new _DSimpleCommandOption(data);
  }
};

// src/decorators/decorators/Bot.ts
var import_internal8 = require("@discordx/internal");
function Bot(...botIds) {
  return function(target, key, descriptor) {
    MetadataStorage.instance.addModifier(
      import_internal8.Modifier.create(
        (original) => {
          original.botIds = [
            ...original.botIds,
            ...botIds.filter((botId) => !original.botIds.includes(botId))
          ];
          if (original instanceof DDiscord) {
            [
              ...original.applicationCommands,
              ...original.simpleCommands,
              ...original.buttons,
              ...original.selectMenus,
              ...original.events
            ].forEach((ob) => {
              ob.botIds = [
                ...ob.botIds,
                ...botIds.filter((botId) => !ob.botIds.includes(botId))
              ];
            });
          }
        },
        DApplicationCommand,
        DSimpleCommand,
        DDiscord,
        DComponent,
        DOn,
        DReaction
      ).decorateUnknown(target, key, descriptor)
    );
  };
}

// src/decorators/decorators/ButtonComponent.ts
function ButtonComponent(options) {
  return function(target, key) {
    const button = DComponent.create({
      botIds: options?.botIds,
      guilds: options?.guilds,
      id: options?.id ?? key,
      type: 0 /* Button */
    }).decorate(target.constructor, key, target[key]);
    MetadataStorage.instance.addComponentButton(button);
  };
}

// src/decorators/decorators/ContextMenu.ts
var import_discord7 = require("discord.js");
function ContextMenu(options) {
  return function(target, key) {
    const applicationCommand = DApplicationCommand.create({
      botIds: options.botIds,
      contexts: options.contexts,
      defaultMemberPermissions: options.defaultMemberPermissions,
      description: "",
      dmPermission: options.dmPermission,
      guilds: options.guilds,
      integrationTypes: options.integrationTypes,
      name: options.name ?? key,
      nameLocalizations: options.nameLocalizations,
      type: options.type
    }).decorate(target.constructor, key, target[key]);
    if (options.type === import_discord7.ApplicationCommandType.Message) {
      MetadataStorage.instance.addApplicationCommandMessage(applicationCommand);
    } else {
      MetadataStorage.instance.addApplicationCommandUser(applicationCommand);
    }
  };
}

// src/decorators/decorators/Discord.ts
function Discord() {
  return function(target) {
    const clazz = target;
    const instance = DDiscord.create(clazz.name).decorate(clazz, clazz.name);
    MetadataStorage.instance.addDiscord(instance);
  };
}

// src/decorators/decorators/Guard.ts
var import_internal9 = require("@discordx/internal");
function Guard(...fns) {
  return function(target, key, descriptor) {
    const guards = fns.map((fn) => {
      return DGuard.create(fn).decorateUnknown(
        target,
        key,
        descriptor
      );
    });
    MetadataStorage.instance.addModifier(
      import_internal9.Modifier.create(
        (original) => {
          original.guards = guards;
        },
        DComponent,
        DApplicationCommand,
        DSimpleCommand,
        DOn,
        DDiscord,
        DReaction
      ).decorateUnknown(target, key, descriptor)
    );
  };
}

// src/decorators/decorators/Guild.ts
var import_internal10 = require("@discordx/internal");
function Guild(...guildIds) {
  return function(target, key, descriptor) {
    MetadataStorage.instance.addModifier(
      import_internal10.Modifier.create(
        (original) => {
          original.guilds = [...original.guilds, ...guildIds];
          if (original instanceof DDiscord) {
            [
              ...original.applicationCommands,
              ...original.simpleCommands,
              ...original.buttons,
              ...original.selectMenus
            ].forEach((obj) => {
              obj.guilds = [...obj.guilds, ...guildIds];
            });
          }
        },
        DApplicationCommand,
        DSimpleCommand,
        DDiscord,
        DComponent,
        DReaction
      ).decorateUnknown(target, key, descriptor)
    );
  };
}

// src/decorators/decorators/ModalComponent.ts
function ModalComponent(options) {
  return function(target, key) {
    const button = DComponent.create({
      botIds: options?.botIds,
      guilds: options?.guilds,
      id: options?.id ?? key,
      type: 2 /* Modal */
    }).decorate(target.constructor, key, target[key]);
    MetadataStorage.instance.addComponentModal(button);
  };
}

// src/decorators/decorators/On.ts
function On(options) {
  return function(target, key, descriptor) {
    const clazz = target;
    const on = DOn.create({
      botIds: options?.botIds,
      event: options?.event ?? key,
      once: false,
      priority: options?.priority,
      rest: false
    }).decorate(clazz.constructor, key, descriptor?.value);
    MetadataStorage.instance.addOn(on);
  };
}
On.rest = function(options) {
  return function(target, key, descriptor) {
    const clazz = target;
    const on = DOn.create({
      botIds: options?.botIds,
      event: options?.event ?? key,
      once: false,
      priority: options?.priority,
      rest: true
    }).decorate(clazz.constructor, key, descriptor?.value);
    MetadataStorage.instance.addOn(on);
  };
};

// src/decorators/decorators/Once.ts
function Once(options) {
  return function(target, key, descriptor) {
    const clazz = target;
    const on = DOn.create({
      botIds: options?.botIds,
      event: options?.event ?? key,
      once: true,
      priority: options?.priority,
      rest: false
    }).decorate(clazz.constructor, key, descriptor.value);
    MetadataStorage.instance.addOn(on);
  };
}
Once.rest = function(options) {
  return function(target, key, descriptor) {
    const clazz = target;
    const on = DOn.create({
      botIds: options?.botIds,
      event: options?.event ?? key,
      once: true,
      priority: options?.priority,
      rest: true
    }).decorate(clazz.constructor, key, descriptor?.value);
    MetadataStorage.instance.addOn(on);
  };
};

// src/decorators/decorators/Reaction.ts
function Reaction(options) {
  return function(target, key) {
    const react = DReaction.create({
      aliases: options?.aliases,
      botIds: options?.botIds,
      description: options?.description,
      directMessage: options?.directMessage,
      emoji: options?.emoji ?? key,
      guilds: options?.guilds,
      partial: options?.partial,
      remove: options?.remove
    }).decorate(target.constructor, key, target[key]);
    MetadataStorage.instance.addReaction(react);
  };
}

// src/decorators/decorators/SelectMenuComponent.ts
function SelectMenuComponent(options) {
  return function(target, key) {
    const button = DComponent.create({
      botIds: options?.botIds,
      guilds: options?.guilds,
      id: options?.id ?? key,
      type: 1 /* SelectMenu */
    }).decorate(target.constructor, key, target[key]);
    MetadataStorage.instance.addComponentSelectMenu(button);
  };
}

// src/decorators/decorators/SimpleCommand.ts
function SimpleCommand(options) {
  return function(target, key) {
    const cmd = DSimpleCommand.create({
      aliases: options?.aliases,
      argSplitter: options?.argSplitter,
      botIds: options?.botIds,
      description: options?.description,
      directMessage: options?.directMessage,
      guilds: options?.guilds,
      name: options?.name ?? key,
      prefix: options?.prefix
    }).decorate(target.constructor, key, target[key]);
    MetadataStorage.instance.addSimpleCommand(cmd);
  };
}

// src/decorators/decorators/SimpleCommandOption.ts
var import_internal11 = require("@discordx/internal");
function SimpleCommandOption(options) {
  return function(target, key, index) {
    const option = DSimpleCommandOption.create(options).decorate(
      target.constructor,
      key,
      target[key],
      target.constructor,
      index
    );
    MetadataStorage.instance.addModifier(
      import_internal11.Modifier.create((original) => {
        original.options = [...original.options, option];
      }, DSimpleCommand).decorate(
        target.constructor,
        key,
        target[key],
        target.constructor,
        index
      )
    );
    MetadataStorage.instance.addSimpleCommandOption(option);
  };
}

// src/decorators/decorators/Slash.ts
var import_discord8 = require("discord.js");
function Slash(options) {
  return function(target, key) {
    const name = options.name ?? key;
    SlashNameValidator(name);
    let applicationCommand;
    if (options instanceof import_discord8.SlashCommandBuilder) {
      if (options.options.length > 0) {
        throw Error(
          "The builder options feature is not supported in discordx."
        );
      }
      applicationCommand = DApplicationCommand.create({
        defaultMemberPermissions: options.default_member_permissions,
        description: options.description,
        descriptionLocalizations: options.description_localizations,
        // eslint-disable-next-line @typescript-eslint/no-deprecated
        dmPermission: options.dm_permission ?? true,
        contexts: options.contexts,
        integrationTypes: options.integration_types,
        name,
        nameLocalizations: options.name_localizations,
        nsfw: options.nsfw,
        type: import_discord8.ApplicationCommandType.ChatInput
      });
    } else {
      applicationCommand = DApplicationCommand.create({
        botIds: options.botIds,
        defaultMemberPermissions: options.defaultMemberPermissions,
        description: options.description,
        descriptionLocalizations: options.descriptionLocalizations,
        dmPermission: options.dmPermission ?? true,
        contexts: options.contexts,
        integrationTypes: options.integrationTypes,
        guilds: options.guilds,
        name,
        nameLocalizations: options.nameLocalizations,
        nsfw: options.nsfw,
        type: import_discord8.ApplicationCommandType.ChatInput
      });
    }
    applicationCommand.decorate(target.constructor, key, target[key]);
    MetadataStorage.instance.addApplicationCommandSlash(applicationCommand);
  };
}

// src/decorators/decorators/SlashChoice.ts
var import_internal12 = require("@discordx/internal");
function SlashChoice(...choices) {
  return function(target, key, index) {
    MetadataStorage.instance.addModifier(
      import_internal12.Modifier.create((original) => {
        const allChoices = choices.map((choice) => {
          const resolveChoice = typeof choice === "number" ? { name: choice.toString(), value: choice } : typeof choice === "string" ? { name: choice, value: choice } : choice;
          return DApplicationCommandOptionChoice.create(resolveChoice);
        });
        original.choices = [...allChoices, ...original.choices];
      }, DApplicationCommandOption).decorate(
        target.constructor,
        key,
        target[key],
        target.constructor,
        index
      )
    );
  };
}

// src/decorators/decorators/SlashGroup.ts
var import_internal13 = require("@discordx/internal");
function SlashGroup(options, root) {
  return function(target, key, descriptor) {
    if (typeof options === "string") {
      MetadataStorage.instance.addModifier(
        import_internal13.Modifier.create(
          (original) => {
            if (original instanceof DDiscord) {
              [...original.applicationCommands].forEach((obj) => {
                obj.group = root ?? options;
                obj.subgroup = root ? options : void 0;
              });
            } else {
              original.group = root ?? options;
              original.subgroup = root ? options : void 0;
            }
          },
          DApplicationCommand,
          DDiscord
        ).decorateUnknown(target, key, descriptor)
      );
    } else {
      SlashNameValidator(options.name);
      const clazz = target;
      if (options.root) {
        MetadataStorage.instance.addApplicationCommandSlashSubGroups(
          DApplicationCommandGroup.create({
            name: options.name,
            payload: {
              description: options.description,
              descriptionLocalizations: options.descriptionLocalizations,
              nameLocalizations: options.nameLocalizations
            },
            root: options.root
          }).decorate(clazz, clazz.name)
        );
      } else {
        MetadataStorage.instance.addApplicationCommandSlashGroups(
          DApplicationCommandGroup.create({
            name: options.name,
            payload: {
              contexts: options.contexts,
              defaultMemberPermissions: options.defaultMemberPermissions,
              description: options.description,
              descriptionLocalizations: options.descriptionLocalizations,
              dmPermission: options.dmPermission,
              integrationTypes: options.integrationTypes,
              nameLocalizations: options.nameLocalizations
            }
          }).decorate(clazz, key ?? clazz.name)
        );
      }
    }
  };
}

// src/decorators/decorators/SlashOption.ts
var import_internal14 = require("@discordx/internal");
var import_discord9 = require("discord.js");
function SlashOption(options, transformer) {
  return function(target, key, index) {
    SlashNameValidator(options.name);
    let option;
    if (options instanceof import_discord9.SlashCommandAttachmentOption || options instanceof import_discord9.SlashCommandBooleanOption || options instanceof import_discord9.SlashCommandRoleOption || options instanceof import_discord9.SlashCommandMentionableOption || options instanceof import_discord9.SlashCommandUserOption) {
      option = DApplicationCommandOption.create({
        description: options.description,
        descriptionLocalizations: options.description_localizations,
        index,
        name: options.name,
        nameLocalizations: options.name_localizations,
        required: options.required,
        transformer,
        type: options.type
      });
    } else if (options instanceof import_discord9.SlashCommandChannelOption) {
      option = DApplicationCommandOption.create({
        channelType: options.channel_types,
        description: options.description,
        descriptionLocalizations: options.description_localizations,
        index,
        name: options.name,
        nameLocalizations: options.name_localizations,
        required: options.required,
        transformer,
        type: options.type
      });
    } else if (options instanceof import_discord9.SlashCommandIntegerOption || options instanceof import_discord9.SlashCommandNumberOption) {
      const choices = options.choices?.map(
        (choice) => DApplicationCommandOptionChoice.create(choice)
      );
      option = DApplicationCommandOption.create({
        autocomplete: options.autocomplete,
        choices,
        description: options.description,
        descriptionLocalizations: options.description_localizations,
        index,
        maxValue: options.max_value,
        minValue: options.min_value,
        name: options.name,
        nameLocalizations: options.name_localizations,
        required: options.required,
        transformer,
        type: options.type
      });
    } else if (options instanceof import_discord9.SlashCommandStringOption) {
      const choices = options.choices?.map(
        (choice) => DApplicationCommandOptionChoice.create(choice)
      );
      option = DApplicationCommandOption.create({
        autocomplete: options.autocomplete,
        choices,
        description: options.description,
        descriptionLocalizations: options.description_localizations,
        index,
        maxLength: options.max_length,
        minLength: options.min_length,
        name: options.name,
        nameLocalizations: options.name_localizations,
        required: options.required,
        transformer,
        type: options.type
      });
    } else {
      option = DApplicationCommandOption.create({
        autocomplete: options.autocomplete,
        channelType: options.channelTypes,
        description: options.description,
        descriptionLocalizations: options.descriptionLocalizations,
        index,
        maxLength: options.maxLength,
        maxValue: options.maxValue,
        minLength: options.minLength,
        minValue: options.minValue,
        name: options.name,
        nameLocalizations: options.nameLocalizations,
        required: options.required,
        transformer,
        type: options.type
      });
    }
    option.decorate(
      target.constructor,
      key,
      target[key],
      target.constructor,
      index
    );
    MetadataStorage.instance.addModifier(
      import_internal14.Modifier.create((original) => {
        original.options = [...original.options, option];
      }, DApplicationCommand).decorate(
        target.constructor,
        key,
        target[key],
        target.constructor,
        index
      )
    );
    MetadataStorage.instance.addApplicationCommandSlashOption(option);
  };
}

// src/logic/metadata/MetadataStorage.ts
var import_di3 = require("@discordx/di");
var import_internal15 = require("@discordx/internal");
var import_discord10 = require("discord.js");
var import_findIndex = __toESM(require("lodash/findIndex.js"));
var MetadataStorage = class _MetadataStorage {
  // internal
  static _isBuilt = false;
  static _instance;
  _discords = [];
  _guards = [];
  _modifiers = [];
  // events
  _events = [];
  // custom Handlers
  _buttonComponents = [];
  _modalComponents = [];
  _selectMenuComponents = [];
  // reactions
  _reactions = [];
  // simple command
  _simpleCommandOptions = [];
  _simpleCommands = [];
  _simpleCommandsByName = [];
  _simpleCommandMappedPrefix = /* @__PURE__ */ new Set();
  // discord commands
  _applicationCommandMessages = [];
  _applicationCommandSlashes = [];
  _applicationCommandSlashesFlat = [];
  _applicationCommandSlashOptions = [];
  _applicationCommandUsers = [];
  // groups
  _applicationCommandSlashGroups = [];
  _applicationCommandSlashSubGroups = [];
  // static getters
  static clear() {
    this._isBuilt = false;
    this._instance = new _MetadataStorage();
  }
  static get isBuilt() {
    return this._isBuilt;
  }
  static get instance() {
    this._instance ??= new _MetadataStorage();
    return this._instance;
  }
  static set instance(value) {
    this._instance = value;
  }
  // getters
  get applicationCommandSlashes() {
    return this._applicationCommandSlashes;
  }
  get applicationCommandSlashesFlat() {
    return this._applicationCommandSlashesFlat;
  }
  get applicationCommandSlashOptions() {
    return this._applicationCommandSlashOptions;
  }
  get applicationCommandSlashGroups() {
    return this._applicationCommandSlashGroups;
  }
  get applicationCommandSlashSubGroups() {
    return this._applicationCommandSlashSubGroups;
  }
  get applicationCommandUsers() {
    return this._applicationCommandUsers;
  }
  get applicationCommandMessages() {
    return this._applicationCommandMessages;
  }
  get applicationCommands() {
    return [
      ...this.applicationCommandSlashes,
      ...this.applicationCommandMessages,
      ...this.applicationCommandUsers
    ];
  }
  get buttonComponents() {
    return this._buttonComponents;
  }
  get discords() {
    return this._discords;
  }
  get discordMembers() {
    return [
      ...this._applicationCommandSlashes,
      ...this._applicationCommandUsers,
      ...this._applicationCommandMessages,
      ...this._reactions,
      ...this._simpleCommands,
      ...this._events,
      ...this._buttonComponents,
      ...this._modalComponents,
      ...this._selectMenuComponents
    ];
  }
  get events() {
    return this._events;
  }
  get isBuilt() {
    return _MetadataStorage._isBuilt;
  }
  get modalComponents() {
    return this._modalComponents;
  }
  get reactions() {
    return this._reactions;
  }
  get selectMenuComponents() {
    return this._selectMenuComponents;
  }
  get simpleCommandsByName() {
    return this._simpleCommandsByName;
  }
  get simpleCommandMappedPrefix() {
    return Array.from(this._simpleCommandMappedPrefix);
  }
  get simpleCommands() {
    return this._simpleCommands;
  }
  /**
   * Get the list of used events without duplications
   */
  get usedEvents() {
    return this.events.reduce((prev, event, index) => {
      const found = this.events.find((event2) => event.event === event2.event);
      const foundIndex = found ? this.events.indexOf(found) : -1;
      if (foundIndex === index || found?.once !== event.once) {
        prev.push(event);
      }
      return prev;
    }, []);
  }
  addApplicationCommandSlash(slash) {
    this._applicationCommandSlashes.push(slash);
  }
  addApplicationCommandUser(slash) {
    this._applicationCommandUsers.push(slash);
  }
  addApplicationCommandMessage(slash) {
    this._applicationCommandMessages.push(slash);
  }
  addApplicationCommandSlashOption(option) {
    this._applicationCommandSlashOptions.push(option);
  }
  addApplicationCommandSlashGroups(group) {
    this._applicationCommandSlashGroups.push(group);
  }
  addApplicationCommandSlashSubGroups(subGroup) {
    this._applicationCommandSlashSubGroups.push(subGroup);
  }
  addComponentButton(button) {
    this._buttonComponents.push(button);
  }
  addComponentModal(selectMenu) {
    this._modalComponents.push(selectMenu);
  }
  addComponentSelectMenu(selectMenu) {
    this._selectMenuComponents.push(selectMenu);
  }
  addDiscord(discord) {
    this._discords.push(discord);
    import_di3.DIService.engine.addService(discord.classRef);
  }
  addGuard(guard) {
    this._guards.push(guard);
    import_di3.DIService.engine.addService(guard.classRef);
  }
  addModifier(modifier) {
    this._modifiers.push(modifier);
  }
  addOn(on) {
    this._events.push(on);
  }
  addReaction(reaction) {
    this._reactions.push(reaction);
  }
  addSimpleCommand(cmd) {
    this._simpleCommands.push(cmd);
  }
  addSimpleCommandOption(cmdOption) {
    this._simpleCommandOptions.push(cmdOption);
  }
  async build() {
    if (_MetadataStorage.isBuilt) {
      return;
    }
    _MetadataStorage._isBuilt = true;
    this.discordMembers.forEach((member) => {
      const discord = this._discords.find((instance) => {
        return instance.from === member.from;
      });
      if (!discord) {
        throw Error(
          `Did you forget to use the @discord decorator on your class: ${String(member.from.name)}
read more at https://discordx.js.org/docs/discordx/decorators/general/discord

`
        );
      }
      member.discord = discord;
      if (member instanceof DApplicationCommand) {
        discord.applicationCommands.push(member);
      }
      if (member instanceof DSimpleCommand) {
        discord.simpleCommands.push(member);
      }
      if (member instanceof DReaction) {
        discord.reactions.push(member);
      }
      if (member instanceof DOn) {
        discord.events.push(member);
      }
      if (member instanceof DComponent) {
        if (member.type === 0 /* Button */) {
          discord.buttons.push(member);
        } else if (member.type === 1 /* SelectMenu */) {
          discord.selectMenus.push(member);
        }
      }
    });
    await import_internal15.Modifier.applyFromModifierListToList(this._modifiers, this._discords);
    await import_internal15.Modifier.applyFromModifierListToList(this._modifiers, this._events);
    await import_internal15.Modifier.applyFromModifierListToList(
      this._modifiers,
      this._applicationCommandSlashes
    );
    await import_internal15.Modifier.applyFromModifierListToList(
      this._modifiers,
      this._applicationCommandSlashOptions
    );
    await import_internal15.Modifier.applyFromModifierListToList(
      this._modifiers,
      this._applicationCommandMessages
    );
    await import_internal15.Modifier.applyFromModifierListToList(
      this._modifiers,
      this._applicationCommandUsers
    );
    await import_internal15.Modifier.applyFromModifierListToList(
      this._modifiers,
      this._simpleCommands
    );
    await import_internal15.Modifier.applyFromModifierListToList(
      this._modifiers,
      this._simpleCommandOptions
    );
    await import_internal15.Modifier.applyFromModifierListToList(
      this._modifiers,
      this._buttonComponents
    );
    await import_internal15.Modifier.applyFromModifierListToList(
      this._modifiers,
      this._modalComponents
    );
    await import_internal15.Modifier.applyFromModifierListToList(
      this._modifiers,
      this._reactions
    );
    await import_internal15.Modifier.applyFromModifierListToList(
      this._modifiers,
      this._selectMenuComponents
    );
    this._applicationCommandSlashesFlat = this._applicationCommandSlashes;
    this._applicationCommandSlashes = this.groupSlashes();
    this.buildSimpleCommands();
    this._events.sort(
      (a, b) => `${a.event}__${a.priority.toString()}` > `${b.event}__${b.priority.toString()}` ? 1 : -1
    );
  }
  buildSimpleCommands() {
    this._simpleCommands.forEach((cmd) => {
      if (cmd.prefix) {
        toStringArray(cmd.prefix).forEach(
          (pfx) => this._simpleCommandMappedPrefix.add(pfx)
        );
      }
      if ((0, import_findIndex.default)(this._simpleCommandsByName, { name: cmd.name }) !== -1) {
        throw Error(`Duplicate simple command name: ${cmd.name}`);
      }
      this._simpleCommandsByName.push({ command: cmd, name: cmd.name });
      cmd.aliases.forEach((alias) => {
        if ((0, import_findIndex.default)(this._simpleCommandsByName, { name: alias }) !== -1) {
          throw Error(`Duplicate simple command name: ${alias}`);
        }
        this._simpleCommandsByName.push({ command: cmd, name: alias });
      });
    });
    this._simpleCommandsByName = this._simpleCommandsByName.sort(
      function(a, b) {
        return b.name.length - a.name.length;
      }
    );
  }
  groupSlashes() {
    const groupedSlashes = /* @__PURE__ */ new Map();
    this._applicationCommandSlashGroups.forEach((group) => {
      if (!group.payload.description) {
        throw Error(`Description required for slash group: ${group.name}`);
      }
      const slashParent = DApplicationCommand.create({
        contexts: group.payload.contexts,
        defaultMemberPermissions: group.payload.defaultMemberPermissions,
        description: group.payload.description,
        descriptionLocalizations: group.payload.descriptionLocalizations,
        dmPermission: group.payload.dmPermission,
        integrationTypes: group.payload.integrationTypes,
        name: group.name,
        nameLocalizations: group.payload.nameLocalizations,
        type: import_discord10.ApplicationCommandType.ChatInput
      }).decorate(group.classRef, group.key, group.method);
      const discord = this._discords.find((instance) => {
        return instance.from === slashParent.from;
      });
      if (!discord) {
        return;
      }
      slashParent.discord = discord;
      slashParent.guilds = [...slashParent.discord.guilds];
      slashParent.botIds = [...slashParent.discord.botIds];
      groupedSlashes.set(group.name, slashParent);
      const slashes = this._applicationCommandSlashes.filter((slash) => {
        return slash.group === slashParent.name && !slash.subgroup;
      });
      slashes.forEach((slash) => {
        slashParent.options.push(slash.toSubCommand());
      });
      this._applicationCommandSlashesFlat.forEach((slash) => {
        if (slash.group === slashParent.name) {
          slash.guilds = slashParent.guilds;
          slash.botIds = slashParent.botIds;
        }
      });
    });
    this._applicationCommandSlashSubGroups.forEach((subGroup) => {
      if (!subGroup.payload.description) {
        throw Error(
          `Description required for slash sub group: ${subGroup.name} (root: ${subGroup.root ?? "unknown"})`
        );
      }
      const option = DApplicationCommandOption.create({
        description: subGroup.payload.description,
        descriptionLocalizations: subGroup.payload.descriptionLocalizations,
        name: subGroup.name,
        nameLocalizations: subGroup.payload.nameLocalizations,
        required: true,
        type: import_discord10.ApplicationCommandOptionType.SubcommandGroup
      }).decorate(subGroup.classRef, subGroup.key, subGroup.method);
      const slashes = this._applicationCommandSlashes.filter((slash) => {
        return slash.group === subGroup.root && slash.subgroup === subGroup.name;
      });
      slashes.forEach((slash) => {
        option.options.push(slash.toSubCommand());
      });
      const groupName = subGroup.root ?? subGroup.name;
      const parentGroup = groupedSlashes.get(groupName);
      if (!parentGroup) {
        throw Error(`A subgroup declared without root: ${groupName}`);
      }
      parentGroup.options.push(option);
    });
    return [
      ...this._applicationCommandSlashes.filter((s) => !s.group && !s.subgroup),
      ...Array.from(groupedSlashes.values())
    ];
  }
  /**
   * Trigger a discord event
   *
   * @param options - Even data
   */
  trigger(options) {
    const eventsToExecute = this._events.filter((on) => {
      return on.event === options.event && on.once === options.once && on.rest === options.rest;
    });
    return async (...params) => {
      const responses = [];
      await Promise.all(
        eventsToExecute.map(async (ev) => {
          if (!ev.isBotAllowed(options.client.botId)) {
            return;
          }
          const res = await ev.execute(options.guards, params, options.client);
          responses.push(res);
        })
      );
      return responses;
    };
  }
};

// src/types/core/literal.ts
var SpecialCharactersList = [
  "~",
  "`",
  "!",
  "@",
  "#",
  "$",
  "%",
  "^",
  "&",
  "*",
  "(",
  ")",
  "+",
  "=",
  "{",
  "}",
  "[",
  "]",
  "|",
  "\\",
  "/",
  ":",
  ";",
  '"',
  "'",
  "<",
  ">",
  ",",
  ".",
  "?",
  " "
];

// src/types/public/enum.ts
var ComponentType = /* @__PURE__ */ ((ComponentType2) => {
  ComponentType2[ComponentType2["Button"] = 0] = "Button";
  ComponentType2[ComponentType2["SelectMenu"] = 1] = "SelectMenu";
  ComponentType2[ComponentType2["Modal"] = 2] = "Modal";
  return ComponentType2;
})(ComponentType || {});
var SimpleCommandParseType = /* @__PURE__ */ ((SimpleCommandParseType2) => {
  SimpleCommandParseType2[SimpleCommandParseType2["notCommand"] = 0] = "notCommand";
  SimpleCommandParseType2[SimpleCommandParseType2["notFound"] = 1] = "notFound";
  return SimpleCommandParseType2;
})(SimpleCommandParseType || {});

// src/types/public/simple command.ts
var SimpleCommandOptionType = /* @__PURE__ */ ((SimpleCommandOptionType2) => {
  SimpleCommandOptionType2[SimpleCommandOptionType2["String"] = 0] = "String";
  SimpleCommandOptionType2[SimpleCommandOptionType2["Number"] = 1] = "Number";
  SimpleCommandOptionType2[SimpleCommandOptionType2["Boolean"] = 2] = "Boolean";
  SimpleCommandOptionType2[SimpleCommandOptionType2["User"] = 3] = "User";
  SimpleCommandOptionType2[SimpleCommandOptionType2["Channel"] = 4] = "Channel";
  SimpleCommandOptionType2[SimpleCommandOptionType2["Role"] = 5] = "Role";
  SimpleCommandOptionType2[SimpleCommandOptionType2["Mentionable"] = 6] = "Mentionable";
  return SimpleCommandOptionType2;
})(SimpleCommandOptionType || {});

// src/util/common.ts
function toStringArray(...input) {
  return input.flatMap((item) => Array.isArray(item) ? item : [item]);
}

// src/util/comparison.ts
var import_isEqual = __toESM(require("lodash/isEqual.js"));
var import_omit = __toESM(require("lodash/omit.js"));
function jsonToString(obj) {
  return JSON.stringify(
    obj,
    (key, value) => typeof value === "bigint" ? value.toString() : value
  );
}
function RecursivelyMatchField(object, keys, onMatch) {
  Object.keys(object).forEach(function(k) {
    if (keys.includes(k)) {
      onMatch(object, k);
    }
    if (object[k] && typeof object[k] === "object") {
      RecursivelyMatchField(object[k], keys, onMatch);
    }
  });
}
function isApplicationCommandEqual(findCommand, DCommand, isGuild) {
  const commandJson = findCommand.toJSON();
  const rawData = DCommand.toJSON();
  RecursivelyMatchField(
    commandJson,
    ["descriptionLocalizations", "nameLocalizations"],
    (object, key) => {
      if (object[key] === void 0) {
        object[key] = null;
      }
    }
  );
  RecursivelyMatchField(
    commandJson,
    ["descriptionLocalized", "nameLocalized", "dmPermission", "nsfw"],
    (object, key) => {
      if (object[key] === null) {
        object[key] = void 0;
      }
    }
  );
  if (isGuild) {
    RecursivelyMatchField(
      rawData,
      ["dmPermission"],
      (object, key) => {
        object[key] = void 0;
      }
    );
  }
  const firstJson = JSON.parse(
    jsonToString(
      (0, import_omit.default)(
        commandJson,
        "id",
        "applicationId",
        "guild",
        "guildId",
        "version",
        "descriptionLocalized",
        "nameLocalized",
        "permissions",
        "defaultPermission"
      )
    )
  );
  const secondJson = JSON.parse(jsonToString(rawData));
  const response = (0, import_isEqual.default)(firstJson, secondJson);
  return response;
}

// src/util/resolve-guilds.ts
var import_isEqual2 = __toESM(require("lodash/isEqual.js"));
var import_uniqWith = __toESM(require("lodash/uniqWith.js"));
var resolveIGuilds = async (client, command, guilds) => {
  const guildX = await Promise.all(
    guilds.map(
      (guild) => typeof guild === "function" ? guild(client, command) : guild
    )
  );
  return (0, import_uniqWith.default)(guildX.flat(1), import_isEqual2.default);
};

// src/util/slash-name-validator.ts
function SlashNameValidator(name) {
  const isNotValid = name.length === 0 || name.length > 32 || name.toLowerCase() !== name || SpecialCharactersList.some((c) => name.includes(c));
  if (isNotValid) {
    throw Error(
      `Invalid slash name: ${name}
Name must only be lowercase with no space as per Discord guidelines (https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-naming)
`
    );
  }
  return true;
}

// src/index.ts
__reExport(index_exports, require("@discordx/di"), module.exports);
__reExport(index_exports, require("@discordx/internal"), module.exports);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ApplicationCommandMixin,
  Bot,
  ButtonComponent,
  Client,
  ComponentType,
  ContextMenu,
  DApplicationCommand,
  DApplicationCommandGroup,
  DApplicationCommandOption,
  DApplicationCommandOptionChoice,
  DComponent,
  DDiscord,
  DGuard,
  DOn,
  DReaction,
  DSimpleCommand,
  DSimpleCommandOption,
  Discord,
  Guard,
  Guild,
  MetadataStorage,
  ModalComponent,
  On,
  Once,
  Reaction,
  RecursivelyMatchField,
  SelectMenuComponent,
  SimpleCommand,
  SimpleCommandMessage,
  SimpleCommandOption,
  SimpleCommandOptionType,
  SimpleCommandParseType,
  Slash,
  SlashChoice,
  SlashGroup,
  SlashNameValidator,
  SlashOption,
  SpecialCharactersList,
  isApplicationCommandEqual,
  resolveIGuilds,
  toStringArray,
  ...require("@discordx/di"),
  ...require("@discordx/internal")
});
